DataFormats/PatCandidates/src/PackedCandidate.cc

```diff
@@ -62,9 +62,12 @@ void pat::PackedCandidate::packVtx(bool unpackAfterwards) {
 }
 
 void pat::PackedCandidate::unpack() const {
-    p4_ = PolarLorentzVector(MiniFloatConverter::float16to32(packedPt_),
+    float pt = MiniFloatConverter::float16to32(packedPt_);
+    double shift = (pt<1. ? 0.1*pt : 0.1/pt); // shift particle phi to break degeneracies in angular separations
+    double phi = int16_t(packedPhi_)*3.2f/std::numeric_limits<int16_t>::max() + shift*3.2/std::numeric_limits<int16_t>::max();
```

Comment:
About the ghosts, my question was mostly about how their phi is defined:
if we use packedPhi_/2^15 + 0.1*pt/2^15
if pt is ~1E-19 (or what's the value?), then it will be below double
precision and ghosts will be on the grid.


On 6/12/15 8:13 AM, Dinko Ferencek wrote:
> In DataFormats/PatCandidates/src/PackedCandidate.cc
> <https://github.com/cms-sw/cmssw/pull/9554#discussion_r32314079>:
> 
>> @@ -62,9 +62,12 @@ void pat::PackedCandidate::packVtx(bool unpackAfterwards) {
>>  }
>>  
>>  void pat::PackedCandidate::unpack() const {
>> -    p4_ = PolarLorentzVector(MiniFloatConverter::float16to32(packedPt_),
>> +    float pt = MiniFloatConverter::float16to32(packedPt_);
>> +    double shift = (pt<1. ? 0.1*pt : 0.1/pt); // shift particle phi to break degeneracies in angular separations
>> +    double phi = int16_t(packedPhi_)*3.2f/std::numeric_limits<int16_t>::max() + shift*3.2/std::numeric_limits<int16_t>::max();
> 
> Since particles are placed on an eta-phi grid, the degeneracy occurs in
> dR separation of particle pairs. In other words, there can be more than
> one pair with the same dR separation and this is particularly
> troublesome for the CA algorithm which uses dR as the distance measure
> (anti-kt and kt break the degenracy by also using particle pt). These
> degeneracies mostly affect early stages of jet clustering so essentially
> jets of any size are affected, not just fat jets with large cone size.
> 
> The ghost association will still work just fine since the particle pt's
> are not changed (0.1*pt, for pt<1, only determines by how much to shift
> the phi value in units of the grid spacing).
> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/9554/files#r32314079>.
> 

-- 
***********************************************************************
 Vyacheslav (Slava) Krutelyov
 TAMU: Physics Dept Texas A&M MS4242, College Station, TX 77843-4242
 CERN: 42-R-027
 AIM/Skype: siava16              googleTalk: slava77@gmail.com
 (630) 291-5128  Cell (US)       +41 76 275 7116 Cell (CERN)

-----------------------------------------------------------------------